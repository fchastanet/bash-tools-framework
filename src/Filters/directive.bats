#!/usr/bin/env bash

# shellcheck source=src/batsHeaders.sh
source "$(cd "${BATS_TEST_DIRNAME}/.." && pwd)/batsHeaders.sh"

# shellcheck source=src/Filters/directive.sh
source "${BATS_TEST_DIRNAME}/directive.sh"

function Filters::directive::keepOnlyDirective { #@test
  run Filters::directive "${FILTER_DIRECTIVE_KEEP_ONLY_HEADERS}" "${BATS_TEST_DIRNAME}/testsData/directive.sh"
  assert_output "$(cat "${BATS_TEST_DIRNAME}/testsData/directive.invert0.expected.txt")"
}

function Filters::directive::removeOnlyDirective { #@test
  run Filters::directive "${FILTER_DIRECTIVE_REMOVE_HEADERS}" "${BATS_TEST_DIRNAME}/testsData/directive.sh"
  assert_output "$(cat "${BATS_TEST_DIRNAME}/testsData/directive.invert1.expected.sh")"
}

function Filters::directive::defaultInvertValue { #@test
  {
    run Filters::directive
    assert_output "$(cat "${BATS_TEST_DIRNAME}/testsData/directive.invert0.expected.txt")"
  } <"${BATS_TEST_DIRNAME}/testsData/directive.sh"
}

function Filters::directive::noDirective { #@test
  run Filters::directive "${FILTER_DIRECTIVE_REMOVE_HEADERS}" "${BATS_TEST_DIRNAME}/testsData/noDirective.sh"
  assert_output "$(cat "${BATS_TEST_DIRNAME}/testsData/noDirective.sh")"
}
