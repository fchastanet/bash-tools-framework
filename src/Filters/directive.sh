#!/usr/bin/env bash

readonly FILTER_DIRECTIVE_KEEP_ONLY_HEADERS="0"
# shellcheck disable=SC2034
readonly FILTER_DIRECTIVE_REMOVE_HEADERS="1"

# filter the directive only at the beginning of the file
# we do not want to filter eventual documentation in the
# rest of the script
# @param {bool} invert $1 Invert the sense of matching,
# to select non-matching lines:
# - 0 to keep only directive
# - 1 to remove directive and keep the rest of the file
# shellcheck disable=SC2120
Filters::directive() {
  local invert="${1:-${FILTER_DIRECTIVE_KEEP_ONLY_HEADERS}}"
  shift || true
  awk -v invert="${invert}" '
    BEGIN {
      write=!invert
      headerProcessed=0
    }
    {
      line = $0
      if (/^# (BIN_FILE=|VAR_[^=]*=|EMBED ).*$/) {
        write=(invert==0) || (headerProcessed==1)
      } else if (/^#!/) {
        write=invert
      } else {
        if (invert == 0) {
          # optimization no need to process more lines
          exit 0
        }
        headerProcessed=1
        write=invert
      }
      if (write==1) {
        print line
      } else {
        next
      }
    }
  ' "$@" || test $? = 1
}
