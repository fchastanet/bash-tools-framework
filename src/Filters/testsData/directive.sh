#!/usr/bin/env bash
# BIN_FILE=<% ${FRAMEWORK_ROOT_DIR} %>/bin/awkLint

.INCLUDE "${ORIGINAL_TEMPLATE_DIR}/_includes/_header.tpl"

test # comment after command

true

# BIN_FILE=<% ${FRAMEWORK_ROOT_DIR} %>/bin/awkLint2

otherTest
