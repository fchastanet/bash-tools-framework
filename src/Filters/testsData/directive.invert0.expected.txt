# BIN_FILE=<% ${ROOT_DIR} %>/bin/awkLint
